crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs

```diff
@@ -0,0 +1,387 @@
+///   2. Assign a DHT arc to each agent such that they overlap with the next agent's start location.
+///   3. Connect each agent to the next agent (circular), so that we know they are aware of each other.
+///   4. Publish an op with a basis location set to the location of the 4th agent. This should also be visible to the 3rd agent by 2. above.
+///   5. Wait for the 3rd to receive the data.
+///   6. Assert that the op was never published to the 1st, 2nd, or 5th agents. (Note that we cannot check if we sent it to ourselves because the op was already in our store)
+#[cfg(feature = "tx5")]
+#[ignore = "This test is flaky, possibly because what it is testing is flaky"]
+#[tokio::test(flavor = "multi_thread")]
+async fn publish_to_basis_from_inside() {
+    holochain_trace::test_run();
+
+    let (bootstrap_addr, _bootstrap_handle) = start_bootstrap().await;
+    let (signal_url, _signal_srv_handle) = start_signal_srv().await;
+
+    let space = Arc::new(fixt!(KitsuneSpace));
+
+    let tuner = |mut params: tuning_params_struct::KitsuneP2pTuningParams| {
+        params.gossip_arc_clamping = "none".to_string();
+        params.gossip_dynamic_arcs = false; // Don't update the arcs dynamically, use the initial value
+        params.disable_recent_gossip = true;
+        params.disable_historical_gossip = true;
+        params
+    };
+
+    // Arcs are this long by default, with an adjustment to ensure overlap.
+    let base_len = u32::MAX / 5;
+
+    let dim = SpaceDimension::standard();
+
+    let mut agents = Vec::new();
+
+    for i in 0..5 {
+        let mut harness = KitsuneTestHarness::try_new("")
+            .await
+            .expect("Failed to setup test harness")
+            .configure_tx5_network(signal_url)
+            .use_bootstrap_server(bootstrap_addr)
+            .update_tuning_params(tuner);
+
+        let sender = harness.spawn().await.expect("should be able to spawn node");
+
+        let mut agent = harness.create_agent().await;
+        let mut found_loc = false;
+        for _ in 0..1000 {
+            let loc = agent.get_loc().as_();
+            if loc > base_len * i && loc < base_len * (i + 1) {
+                found_loc = true;
+                break;
+            }
+
+            // If we didn't find a location in the right range, try again
+            agent = harness.create_agent().await;
+        }
+
+        assert!(
+            found_loc,
+            "Failed to find a location in the right range after 1000 tries"
+        );
+
+        // Distance to the end of the segment, plus the length of the next segment. Guaranteed to
+        // overlap with the next agent and not the one after that.
+        // Because of arc quantisation, the layout won't be perfect, but we can expect overlap at
+        // the start of the agent's arc, with the previous agent.
+        let len =
+            DhtLocation::new(base_len * (i + 1)) - agent.get_loc() + DhtLocation::new(base_len);
```

Comment:
It probably isn't the cause of your flaky-ness because of the extreme unlikelyhood of randomly hitting it, but I'm a little worried about an off-by-one error, especially on the wrapping around the beginning if the section 5 node is at the extreme beginning of its range, and the section zero node is at the extreme end.

Reply:
Totally agree, this isn't accurate enough - it might make sense to grow the arcs a bit more and then have the assertions be dynamically checking that the agents whose arcs actually hold the op are getting it. Rather than hard-coding. But for now, I'm checking that agents whose arcs (I think?) aren't including the test basis aren't getting the ops.. but they are. Very confused :)